diff --git a/react-intl.iife.js b/react-intl.iife.js
index 7efd1f820eb10e32b4d0bad6dfae7d906f1697df..0304a79b7f2175339300db231f69203b3ce0d259 100644
--- a/react-intl.iife.js
+++ b/react-intl.iife.js
@@ -9678,10 +9678,9 @@ For more details see https://formatjs.github.io/docs/getting-started/message-dis
     textComponent: React.Fragment
   });
   var toKeyedReactNodeArray = (children) => {
-    const childrenArray = React.Children.toArray(children);
-    return childrenArray.map((child, index) => {
-      if (React.isValidElement(child)) {
-        return React.createElement(React.Fragment, { key: index }, child);
+    return React.Children.map(children, child => {
+      if (React.isValidElement(child) && child._store) {
+        child._store.validated = 1;
       }
       return child;
     });
diff --git a/src/utils.js b/src/utils.js
index dd435f63b83ec6ec512d8df5d2ae487ddfbd3eff..63dbd08abd74d0ef80b23786cde5425c6bfdfb28 100644
--- a/src/utils.js
+++ b/src/utils.js
@@ -26,12 +26,9 @@ export const DEFAULT_INTL_CONFIG = {
 *
 */
 export const toKeyedReactNodeArray = (children) => {
-	const childrenArray = React.Children.toArray(children);
-	return childrenArray.map((child, index) => {
-		// For React elements, wrap in a keyed Fragment
-		// This creates a new element with a key rather than trying to add one after creation
-		if (React.isValidElement(child)) {
-			return /* @__PURE__ */ _jsx(React.Fragment, { children: child }, index);
+	return React.Children.map(children, child => {
+		if (React.isValidElement(child) && child._store) {
+			child._store.validated = 1;
 		}
 		return child;
 	});
