diff --git a/index.js b/index.js
index 9fb3ce32857ce99211ae7c6d0d6c08b38915364f..4ad4b2e77881da43048675439737bdc4b72e93d2 100644
--- a/index.js
+++ b/index.js
@@ -26,7 +26,7 @@ const DEFAULT_INTL_CONFIG$1 = {
 */
 const toKeyedReactNodeArray = (children) => {
 	return React.Children.toArray(children).map((child, index) => {
-		if (React.isValidElement(child)) return /* @__PURE__ */ jsx(React.Fragment, { children: child }, index);
+		if (React.isValidElement(child) && child._store) child._store.validated = 1;
 		return child;
 	});
 };
diff --git a/react-intl.iife.js b/react-intl.iife.js
index c70d5b62ce043b8ce6a21a62a65e41c0b9b4f75e..f7ab74d16da58990cda539f564e088bddeb77087 100644
--- a/react-intl.iife.js
+++ b/react-intl.iife.js
@@ -3310,7 +3310,7 @@ const DEFAULT_INTL_CONFIG = {
 */
 const toKeyedReactNodeArray = (children) => {
 	return React.Children.toArray(children).map((child, index) => {
-		if (React.isValidElement(child)) return /* @__PURE__ */ jsx(React.Fragment, { children: child }, index);
+		if (React.isValidElement(child) && child._store) child._store.validated = 1;
 		return child;
 	});
 };
diff --git a/server.js b/server.js
index 3fe32f9aa6ccb14ae2f9ab66ae72363bc605b5d7..3cab55692bbd110ae86dd06eee6347fbd14f3a4a 100644
--- a/server.js
+++ b/server.js
@@ -19,7 +19,7 @@ const DEFAULT_INTL_CONFIG$1 = {
 */
 const toKeyedReactNodeArray = (children) => {
 	return React.Children.toArray(children).map((child, index) => {
-		if (React.isValidElement(child)) return /* @__PURE__ */ jsx(React.Fragment, { children: child }, index);
+		if (React.isValidElement(child) && child._store) child._store.validated = 1;
 		return child;
 	});
 };
